src/Transform/ONNX/DecomposeEinsum.cpp

```diff
@@ -0,0 +1,591 @@
+    // with dim 1 for all the other axes. For instance, if output.shape is
+    // (2,4,4,5,4) and axes is [1,2,4] then maskShape is (1,4,4,1,4).
+    // The maskValues has true on the "diagonal", i.e.
+    // maskShape[_,i,i,_,i]==true for i in {0,1,2,3} and false elsewhere.
+    Shape maskShape;
+    for (size_t i = 0; i < output.size(); ++i) {
+      maskShape.push_back(output.subscripts[i] == subscript ? d : 1);
+    }
+    int64_t size = ShapedType::getNumElements(maskShape);
+    SmallVector<bool> maskValues(size, false);
+    // In the flat maskValues representation of mask the true values are evenly
+    // spaced out between maskValues[0]==true,...,maskValues[size-1]==true.
+    assert((size - 1) % (d - 1) == 0);
+    auto distance = (size - 1) / (d - 1);
+    for (int64_t i = 0; i < d; ++i) {
+      maskValues[i * distance] = true;
+    }
+    Value mask = tensor<bool>(maskShape, maskValues, builder.getI1Type());
+    output.value = builder
+                       .create<ONNXWhereOp>(loc, output.type(elementType), mask,
+                           output.value, zeroScalar(elementType))
+                       .getResult();
+    sum(output, axes.drop_front());
+  }
+
+  void diagonalize(Output &output) {
+    auto dups = output.duplicates();
+    for (char x : dups) {
+      Axes axes;
+      for (size_t a = 0; a < output.size(); ++a) {
+        if (output.subscripts[a] == x)
+          axes.push_back(a);
+      }
+      diagonal(output, axes);
+    }
+  }
+
+  void reduce(Output &output, const std::unordered_set<char> &keep) {
+    Axes axes;
+    for (size_t a = 0; a < output.size(); ++a) {
+      if (keep.count(output.subscripts[a]) == 0)
+        axes.push_back(a);
+    }
+    sum(output, axes);
+  }
+
+  std::unordered_set<char> otherSubscripts(
+      const std::vector<einsum::Parameter *> &ignore) const {
+    std::unordered_set<char> subscriptsSet;
+    for (const Output &output : outputs) {
+      if (std::find(ignore.begin(), ignore.end(), &output) == ignore.end())
+        subscriptsSet.insert(
+            output.subscripts.begin(), output.subscripts.end());
+    }
+    if (std::find(ignore.begin(), ignore.end(), &result) == ignore.end())
+      subscriptsSet.insert(result.subscripts.begin(), result.subscripts.end());
+    return subscriptsSet;
+  }
+
+  void transpose(Output &output, const Subscripts &transposedSubscripts) {
+    assert(output.subscripts.size() == transposedSubscripts.size());
+    if (output.subscripts == transposedSubscripts)
+      return;
+
+    Axes perm = transposePerm(output.subscripts, transposedSubscripts);
+    output.subscripts = transposedSubscripts;
+    output.shape = shapePermute(output.shape, perm);
+    output.value = builder
+                       .create<ONNXTransposeOp>(loc, output.type(elementType),
+                           output.value, builder.getI64ArrayAttr(perm))
+                       .getResult();
```

Comment:
We are recently moving to use DialectBuilder instead of building an op directly, which is easier to maintain our code, e.g.:
```
MultiDialectBuilder<OnnxBuilder> create(buider, loc);
create.onnx.transpose(...)
```
The DialectBuilder for ONNX is here: https://github.com/onnx/onnx-mlir/blob/main/src/Dialect/ONNX/DialectBuilder.hpp. Could you please use DialectBuilder instead? and feel free to add new operations if they do not exist.

Reply:
> Could you please use DialectBuilder instead? and feel free to add new operations if they do not exist.

OK. So far I made use of DialectBuilder for the ops that it defines. I'll try to add ops I need that don't exist, e.g. Squeeze, ReduceSum.

I ran into a problem with matmul() which returns toMemRef() which breaks things for me. Should I add a variant of matmul() that omits toMemRef()?

Reply:
>  Should I add a variant of matmul() that omits toMemRef()?

Yes, please. I think it is a worth variant. Thanks!

Reply:
Since the call to toMemref() seemed non-standard, I renamed the old matmul() with toMemref() and changed the existing callers, and then I made plain matmul() not call toMemref() and put that to use in DecomposeEinsum:
https://github.com/onnx/onnx-mlir/pull/1537/commits/9fd808927d0f4ed818ca9408b146f67507a96b76

I'm happy to change it back of course if you wish, just let me know.

Reply:
How about explicitly calling `toMemref` when needed, and removing `matmulToMemref` e.g.
```
Value XtWT =
      create.onnx.toMemref(create.onnx.matmul(matrixAllGatesType, Xt, weightPack.WT));
```

Reply:
> How about explicitly calling toMemref when needed, and removing matmulToMemref

good idea, done

Reply:
Thank you so much!